### PR TITLE
Add a migration to move us from RequestInvite/Trust flow to CreateMember/Verify.

### PR DESCRIPTION
**Pull request by rahulgi**

### CODE REVIEW

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+ * updating their op_code and updating the field called data.to_uid to data.torequest_invite_from_member_id.
+ *
+ * A Trust operation from the inviting member to the new invited member previously indicated that the
+ * new member was legitimate. Now, an explicit Verify operation is required. For every Trust operation
+ * between an inviting member and the person they invited, we create a new Verify operation. The Verify
+ * operation points to the new member's invite video as the Verification video, since all members created
+ * via the RequestInvite flow joined with joint invite videos where both the inviter and the invitee state
+ * their name.
+ *
+ * This migration is idempotent allowing it to be run again if new members go through the RequestInvite flow
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: wooot

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+ * (as could continue to happen until the web-based flow is shut down).
+ *
+ * Note that this migration also modifies the GENESIS RequestInvite and Trust operations. These are the
+ * operations that led to the spontaneous creation and trusting of the GENESIS members (Mark and family).
+ *
+ * It adds the "GENESIS" special value to the request_invite_from_member_id field of the new GENESIS CreateMember
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: Also woooo

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        await operation.ref.update(updatesToOperation);
+        ++updatedOperations;
+      }
+    }
+    if (operationData.op_code === OperationType.TRUST) {
+      const opCreatorUid = operationData.creator_uid;
+
+      const trusterMemberId =
+        opCreatorUid === null &&
+        ORIGINAL_GENESIS_TRUST_OPS.indexOf(operation.id) >= 0
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: can just use `ORIGINAL_GENESIS_TRUST_OPS.includes(operation.id)` but whatevs

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        "data.request_invite_from_member_id": inviterMemberId,
+        "data.to_uid": admin.firestore.FieldValue.delete()
+      };
+      console.log("=============================");
+      console.log("Updating operation", operation.id);
+      console.log("Operation was", operationData);
+      console.log("Update", updatesToOperation);
+      console.log("=============================");
+      if (!dryRun) {
+        await operation.ref.update(updatesToOperation);
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: Should we be doing this all in a transaction?

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        }
+      }
+    }
+  }
+  console.log(`Created ${createdOperations} operations.`);
+  console.log(`Updated ${updatedOperations} operations.`);
+  if (newGenesisVerifyOperations.length > 0) {
+    console.log("!!!!!! IMPORTANT !!!!!!!");
+    console.log(
+      "New GENESIS Verify operations were created, and the original GENESIS Trust operations were deleted."
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: Only if not dry run, right?

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  );
+  const args = process.argv;
+  if (args.length < 4) {
+    console.warn(
+      "Invalid number of arguments. See usage information above. Exiting."
+    );
+    return;
+  }
+  const db = getDb(args[2], args[3]);
+  let isDryRun = args.length > 4 ? args[4] !== "false" : true;
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: const?

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+          ? GENESIS
+          : opCreatorUid;
+      const trustedMemberId = operationData.data.to_uid;
+
+      if (invitedBy[trustedMemberId] === trusterMemberId) {
+        // Check if this is a request-invite confirming Trust operation
+        // Check if the requisite "Verify" operation already exists.
+        // Allows this migration to be idempotent.
+        if (
+          !(await operationsCollection
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: just  want to confirm that querying as we're editing the collection is a safe thing to do. might be better to just keep in memory which verify operations we've created and then flush them at once.

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts (hunk 3 above)
Comment (rahulgi) [reply migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: #oldschoolcoolschool

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts (hunk 5 above)
Comment (rahulgi) [reply migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: correct. this array should only be populated if dryRun is false.

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+  const invitedBy: { [memberId: string]: string } = {};
+
+  let createdOperations = 0;
+  let updatedOperations = 0;
+
+  const newGenesisVerifyOperations: string[] = [];
+
+  for (const i in operations) {
+    await db.runTransaction(async transaction => {
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: Why is it one transaction per operation instead one transaction for the whole migration? This allows for partial application of the migration, the transaction is pointless

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    .slice(2)
+    .toString();
+}
+
+function getInviteVideoUrlForMemberId(memberId: string) {
+  return `https://storage.googleapis.com/raha-video/${memberId}/invite.mp4`;
+}
+
+async function migrateToCreateMemberAndVerify(db: Firestore, dryRun: boolean) {
+  const operationsCollection = db.collection("operations");
```
Comment (osdiab) [inline migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: Use the transaction to get the operations, to ensure it's consistent with the state of the db when the transaction occurs

File: migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts (hunk 5 above)
Comment (osdiab) [reply migrations/2018-08-20-00-migrate-to-create-member-and-verify.ts]: ding

**Pull request merged**